packit/patches.py

```diff
@@ -232,6 +244,74 @@ def linearize_history(git_ref: str):
             env={"FILTER_BRANCH_SQUELCH_WARNING": "1"},
         )
 
+    @staticmethod
+    def process_patches(patches: Dict[str, bytes], commits: List[git.Commit]):
+        """
+        Pair commits (in a source-git repo) with a list patches generated with git-format-patch.
+
+        Pairing is done using commit.hexsha (which is always present in the patch file).
+
+        patch_list (provided List) is then mutated by appending PatchMetadata using
+        the paired information: commit and a path to the patch file.
+        """
+        patch_list: List[PatchMetadata] = []
+        for commit in commits:
+            for patch_name, patch_content in patches.items():
+                # `git format-patch` usually creates one patch for a merge commit,
+                # so some commits won't be covered by a dedicated patch file
+                if commit.hexsha.encode() in patch_content:
+                    path = Path(patch_name)
+                    patch_metadata = PatchMetadata.from_commit(
+                        commit=commit, patch_path=path
+                    )
+
+                    if patch_metadata.ignore:
+                        logger.debug(
+                            f"[IGNORED: {patch_metadata.name}] {commit.summary}"
+                        )
+                    else:
+                        logger.debug(f"[{patch_metadata.name}] {commit.summary}")
+                        patch_list.append(patch_metadata)
+                    break
+        return patch_list
+
+    @staticmethod
+    def process_git_am_style_patches(patch_list: List[PatchMetadata],):
+        """
+        When using `%autosetup -S git_am`, there is a case
+        where a single patch file contains multiple commits.
+        This is problematic for us since only the HEAD patch commit
+        is annotated.
+
+        In this case, we need to:
+        1. detect this is happening - top commit has `squash_commits=True`
+        2. process every commit, reversed (patches[-1] is HEAD)
+        3. append commits to a single patch until `squash_commits=True`
+        4. return new patch list
+        """
+        top_commit = patch_list[-1]
+        if not top_commit.squash_commits:
+            logger.debug(
+                "top commit is squash_commits=False, not the git-am style of patches"
+            )
+            return patch_list
+
+        new_patch_list: List[PatchMetadata] = []
+
+        for patch in reversed(patch_list):
+            if patch.squash_commits:
```

Comment:
@lachmanfrantisek this is the line, so what should I put here?

Reply:
> right now it's inly the squash_metadata=True commit
> Franta, I don't understand, how do you expect me to make the if condition then?
if not current_commit.present_in_specfile?

Can we determine from `PatchMetadata` if we got commit **with** or **without** metadata?

- If yes, we can use that info in the condition, but it looks like we can't:

https://github.com/packit/packit/blob/b5da23c0fa3f205537551b9ed212d8f77d00d705/packit/patches.py#L109-L137

- We can maybe have some `metadata_in_commit: bool` attribute and fill it in the static method above. Will this work?

----

Maybe I am thinking about that too much..;)

Reply:
> Maybe I am thinking about that too much..;)

Yes, I'd love to understand what exactly we are trying to solve here.

Reply:
With your setup, you are squashing all commits between two `squash_metadata=True`, but what we probably want to do is merge all commits without metadata to the one with `squash_metadata=True`.

If we do that, it will be easier to understand by users. Look at this example:

1. `aaaa` with metadata
2. `bbbb` `squash_metadata=True`
3. `cccc` without metadata (needs squash to `dddd`)
4. `dddd` `squash_metadata=True`
5. `eeee` without metadata  (needs squash to `ffff`)
6. `ffff` `squash_metadata=True`
7. `gggg` with metadata

- Users might not know:
  - What will be done with commits `aaaa` and `gggg`?
  - Why we have `squash_metadata=True` in `bbbb` if we don't squash this commit at all?

If you use presence/absence of metadata in your conditional, you can have this instead:

1. `aaaa` with metadata
2. `bbbb` with metadata
3. `cccc` without metadata
4. `dddd` `squash_metadata=True`
5. `eeee` without metadata
6. `ffff` `squash_metadata=True`
7. `gggg` with metadata

And you easily know that `cccc` will be squashed with `dddd` and `eeee` with `ffff`.

Reply:
> With your setup, you are squashing all commits between two `squash_metadata=True`, but what we probably want to do is merge all commits without metadata to the one with `squash_metadata=True`.
> 
> If we do that, it will be easier to understand by users. Look at this example:
> 
>     1. `aaaa` with metadata
> 
>     2. `bbbb` `squash_metadata=True`
> 
>     3. `cccc` without metadata (needs squash to `dddd`)
> 
>     4. `dddd` `squash_metadata=True`
> 
>     5. `eeee` without metadata  (needs squash to `ffff`)
> 
>     6. `ffff` `squash_metadata=True`
> 
>     7. `gggg` with metadata

Who would create such history? Users shouldn't care about this at all - we are solving this problem only for packagers who use git-am patches and this history should exclusively by created only with our tooling. No human should ever write patch metadata in the commit messages.

>     * Users might not know:
>       
>       * What will be done with commits `aaaa` and `gggg`?

I don't know either and certainly don't wanna support such use case.

>       * Why we have `squash_metadata=True` in `bbbb` if we don't squash this commit at all?

As I said, if this happens, we should error out and say that we don't support this and there is an error in our tooling.

> If you use presence/absence of metadata in your conditional, you can have this instead:
> 
>     1. `aaaa` with metadata
> 
>     2. `bbbb` with metadata
> 
>     3. `cccc` without metadata
> 
>     4. `dddd` `squash_metadata=True`
> 
>     5. `eeee` without metadata
> 
>     6. `ffff` `squash_metadata=True`
> 
>     7. `gggg` with metadata
> 
> 
> And you easily know that `cccc` will be squashed with `dddd` and `eeee` with `ffff`.

Honestly, this looks like nightmare to debug and we shouldn't support this.

Let's talk at arch about this.

Reply:
> Honestly, this looks like nightmare to debug and we shouldn't support this.

This is not about the support of various scenarios, but about the readiness of the source-git commits and it's metadata (also by us).

> Let's talk at arch about this.

:+1: Yes, tha would be nice.